### PR TITLE
New resource: aws_lambda_invocation

### LINKED ISSUE #4746: execute aws_lambda_invocation only once
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

After creating an aws_lambda_function I want directly to trigger it for a first initialization. Currently I'm doing this using a provisioner and and some shell commands, which are invoking the the lambda with the AWS CLI.

I've seen there is now an aws_lambda_invocation data source, which does exactly the same like my provisioner. In my case, the disadvantage of a data source is, that the lambda is executed every time I do 
a "terraform plan" or "terraform apply", which is not needed and increases the execution time (and AWS costs).

It would be nice to be able to trigger a lambda function only once

### New or Affected Resource(s)
Option 1
Create:
```hcl
resource "aws_lambda_invocation" "invoke"
```
Option 2
Update:
```hcl
data "aws_lambda_invocation" "invoke"
```
### Potential Terraform Configuration

I would prefer the first option, to create a new ressource type. It would look like the data source:
```hcl
resource "aws_lambda_invocation" "example" {
  function_name = "${aws_lambda_function.lambda_function_test.function_name}"

  input = <<JSON
  {
    "key1": "value1",
    "key2": "value2"
  }
  JSON
}
```

As an alternative an optional attribute in the data source can manage the behavior:
```hcl
data "aws_lambda_invocation" "example" {
  function_name = "${aws_lambda_function.lambda_function_test.function_name}"
  trigger_once  = true

  input = <<JSON
  {
    "key1": "value1",
    "key2": "value2"
  }
  JSON
}
```

### References

No references

### PR DESCRIPTION
**Pull request by dedunumax**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #4746

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
New resource: aws_lambda_invocation (#14388)
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccResourceAwsLambdaInvocation_basic'

...
```

My AWS account permissions are limited. It would be great someone can help me with the acceptance test. Tested functionality locally.

### CODE REVIEW

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by kkotowick** [comment]
Agreed, the data version of this has major limitations and having a resource for lambda invokation would be extremely beneficial. @bflad @YakDriver any chance we can get a review on this? 43 thumbs up on the issue that it closes.

**Comment by geofflancaster** [comment]
@dedunumax per the bot comment above, you'll need to revert the changes to `website/aws.erb`. That may help this get pushed through more quickly.

File: aws/resource_aws_lambda_invocation.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsJSON,
+			},
+
+			"invoke_on_update": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
```
Comment (DJAlPee) [inline aws/resource_aws_lambda_invocation.go:43]: Using `triggers` with a map might be more consistent with other terraform (AWS) resources (and more powerful):
https://github.com/hashicorp/terraform-provider-aws/blob/5283e38ff9fa019c0d8aa3059a41a03cf431a2c2/aws/resource_aws_api_gateway_deployment.go#L45-L50

If I got it right, there would be no further checking (like in #L89) necessary anymore.

**Comment by dedunumax** [comment]
Thanks a lot @pascal236 for fixing the lint. Closing the PR. #19488 has fixes all the issues in this PR.

**Pull request closed without merging**